### ISSUE TITLE
Add config option for timeout behavior

### ISSUE THREAD

**Issue by dvirsky** [RedisLabsModules/RediSearch#294]
Right now if a query surpasses the timeout (500ms), we just abort it and return what we have as a best effort. For some cases this is the correct behaviour, for others it's better to return an error (esp. for aggregations). We should let the user control it. Also it's worth having the option to override the timeout for specific long queries in the query API itself.

See #293 for why the current behaviour is weird and confusing.

**Comment by mnunberg** [comment]
I have in mind something like `[ON_TIMEOUT {fail|return}] ... [TIMEOUT {n}]`.

Another issue here is in cases where several rows have already been emitted; in which case we need to to somehow signal an error in-band, rather than simply abort with an error. I don't believe the Redis protocol supports this, does it?

**Comment by dvirsky** [comment]
no it doesn't. but unless the user requested like a million rows, it probably won't happen. we accumulate the results in the sorting processor, and only after everything finishes we pull the results one by one and and reply. a timeout will always happen while we accumulate the results. I think we don't even check the timeout at a higher level in the pipeline (serialize -> page -> sort -> rank -> search). So so it might be 100% safe to assume that we either time out or get the results. Needs checking in the code but we might be fine.

**Comment by dvirsky** [comment]
BTW when the config gets complicated we can convert it to use CmdParse as well

**Comment by mnunberg** [comment]
TODO: need to add this on a per-query basis.
